lib/services/unified_resource.go

```diff
@@ -0,0 +1,515 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package services
+
+import (
+	"bytes"
+	"context"
+	"sync"
+
+	"github.com/google/btree"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+type UnifiedResourceCacheConfig struct {
+	// Context is a context for opening the
+	// database
+	Context context.Context
+	// BTreeDegree is a degree of B-Tree, 2 for example, will create a
+	// 2-3-4 tree (each node contains 1-3 items and 2-4 children).
+	BTreeDegree int
+	// Clock is a clock for time-related operations
+	Clock clockwork.Clock
+	// Component is a logging component
+	Component string
+	ResourceWatcherConfig
+	ResourceGetter
+}
+
+// UnifiedResourceCache contains a representation of all resources that are displayable in the UI
+type UnifiedResourceCache struct {
+	mu  sync.Mutex
+	log *log.Entry
+	cfg UnifiedResourceCacheConfig
+	// tree is a BTree with items
+	tree            *btree.BTreeG[*Item]
+	initializationC chan struct{}
+	stale           bool
+	once            sync.Once
+	ResourceGetter
+}
+
+// NewUnifiedResourceCache creates a new memory cache that holds the unified resources
+func NewUnifiedResourceCache(ctx context.Context, cfg UnifiedResourceCacheConfig) (*UnifiedResourceCache, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err, "setting defaults for unified resource cache")
+	}
+
+	m := &UnifiedResourceCache{
+		mu: sync.Mutex{},
+		log: log.WithFields(log.Fields{
+			trace.Component: teleport.ComponentMemory,
+		}),
+		cfg: cfg,
+		tree: btree.NewG(cfg.BTreeDegree, func(a, b *Item) bool {
+			return a.Less(b)
+		}),
+		initializationC: make(chan struct{}),
+		ResourceGetter:  cfg.ResourceGetter,
+	}
+
+	err := newWatcher(ctx, m, cfg.ResourceWatcherConfig)
```

Comment:
the watcher requires the cache to be created _before_ we init it since the cache is an argument for the watcher (its the collector used in `newResourceWatcher`). I might be mistaken. 

Other examples I've seen have the watcher init first, but the collector inside of their newwatcher method. So I think one has to come before the other which ever way we do it. I chose this way so that I'd be returning a cache to set on the authServer instead of the watcher like before.